qucs-core/src/components/taperedline.cpp

```diff
@@ -0,0 +1,268 @@
+     nr_double_t Zaux = Z2;
+     Z2 = Z1;
+     Z1 = Zaux;
+  } 
+
+  nr_double_t gamma_max = getPropertyDouble ("Gamma_max");;//Maximum ripple (Klopfenstein weighting only)
+  nr_double_t lstep = L/Nsteps; //Size of the differential elements
+
+  int idx=0;
+  nr_double_t l = lstep/2.0;
+  for (idx = 0, l = lstep/2.0 ; idx < Nsteps; idx++, l += lstep)
+  {
+    // The line is discretized in finite elements. The size of these elements can be considered a differential
+    // length since the impedance change across the actual section is small. 
+    // Taking into account the cascading property of the ABCD matrix, the overall
+    // ABCD matrix can be calculated as the product of the individual ABCD matrices.   
+    if (!strcmp (getPropertyString ("Weighting"), "Exponential"))
+    {
+       Zprofile[idx] = calcExponential(l, L, Z1, Z2);
+    }
+    else
+    {
+        if (!strcmp (getPropertyString ("Weighting"), "Linear"))
+        {
+            Zprofile[idx] = calcLinear(l, L, Z1, Z2);
+        }
+        else
+        {
+           if (!strcmp (getPropertyString ("Weighting"), "Triangular"))
+           {
+              Zprofile[idx] = calcTriangular(l, L, Z1, Z2);
+           }
+           else//Klopfenstein
+           {
+              Zprofile[idx] = calcKlopfenstein(l, L, Z1, Z2, gamma_max);
+           }
+        }
+    }
+ }
+}
+
+//------------------------------------------------------------------
+// Auxiliar function for Klopfenstein profile calculation
+// The recursive calculation algorithm is from
+//   Grossberg, M. A., "Extremely rapid computation of the Klopfenstein
+//   impedance taper," IEEE Proc., vol.56, no.9, pp.1629-1630, Sept. 1968
+//
+// In practice less than 10 iterations are needed in most cases
+nr_double_t taperedline::phi(nr_double_t x, nr_double_t A)
+{
+  nr_double_t ak = 1.0;
+  nr_double_t bk = x *.5;
+  nr_double_t ck = bk;
+  nr_double_t phi = bk;
+  for (int k = 1; k < 20; k++) {
+    ck *= (1.0 - x*x);
+    bk = (ck + 2.0*k*bk) / (2.0*k + 1.0);
+    ak = A*A / (4.0*k*(k + 1.0)) * ak;
+    phi += ak*bk;
+    if (abs(ak*bk) < 1e-7) break;
+  }
+  return phi;
+}
+
+void taperedline::calcSP (nr_double_t frequency) {
+  calcABCDparams(frequency);
+  matrix Stmp = qucs::atos(ABCD, z0, z0);
+  setMatrixS(Stmp);
+}
+
+void taperedline::calcNoiseSP (nr_double_t) {
+  nr_double_t l = getPropertyDouble ("L");
+  if (l < 0) return;
+  // calculate noise using Bosma's theorem
+  nr_double_t T = getPropertyDouble ("Temp");
+  matrix s = getMatrixS ();
+  matrix e = eye (getSize ());
+  setMatrixN (celsius2kelvin (T) / T0 * (e - s * transpose (conj (s))));
+}
+
+void taperedline::calcNoiseAC (nr_double_t) {
+ nr_double_t l = getPropertyDouble ("L");
+  if (l < 0) return;
+  // calculate noise using Bosma's theorem
+  nr_double_t T = getPropertyDouble ("Temp");
+  setMatrixN (4.0 * celsius2kelvin (T) / T0 * real (getMatrixY ()));
+}
+
+void taperedline::initDC (void) {
+  allocMatrixMNA ();
+  // open circuit
```

Comment:
oh, should not really be an open.

Should probably be done like here https://github.com/Qucs/qucs/blob/master/qucs-core/src/components/tline.cpp#L83-L102 . Slightly more difficult since the line is non-uniform.